.github/workflows/release.yml

```diff
@@ -0,0 +1,51 @@
+name: Release
+
+on:
+  push:
+    branches:
+      - master
+      - "[0-9]+" # Major version only
+      - "[0-9]+.[0-9]+" # Major and minor version
+      - "[0-9]+.[0-9]+.[0-9]+" # Major, minor, and patch version
+    tags:
+      - "*"
+
+jobs:
+  release:
+    name: release
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        image_type: [pytorch]
+    env:
+      DOCKER_BUILDKIT: 1
+      IMAGE_TYPE: ${{ matrix.image_type }}
+      QUAY_USER: ${{ secrets.QUAY_RASTERVISION_USER }}
+      QUAY_PASSWORD: ${{ secrets.QUAY_RASTERVISION_PASSWORD }}
+    steps:
+      - uses: actions/checkout@v2
+
+      - run: ./scripts/cibuild
+
+      - uses: codecov/codecov-action@v2
+
+      - run: |
+          echo "GIT_BRANCH=${GITHUB_REF##*/}" >> $GITHUB_ENV
+          echo "GIT_COMMIT=${GITHUB_SHA:0:7}" >> $GITHUB_ENV
```

Comment:
This is the approach [recommended by GitHub Actions](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable) to add environment variables. The `GIT_BRANCH` variable will snag the branch name or tag, and the `GIT_COMMIT` variable gets the short commit hash for image tagging.

Reply:
Consider allowing `GIT_COMMIT` to be sourced by the usual:

```bash
if [[ -n "${GIT_COMMIT}" ]]; then
    GIT_COMMIT="${GIT_COMMIT:0:7}"
else
    GIT_COMMIT="$(git rev-parse --short HEAD)"
fi
```

This works on GitHub Actions, since the Ubuntu runner has `git` installed. Also, consider determining the branch name in a similar fashion.

Reply:
The catch here is that I'd like for it to be available to the workflow itself, and then have it passed along to `cipublish` as the `IMAGE_VERSION` variable. That lets us condition what kind of tag we want to push in the workflow instead of needing to pass the ref into `cipublish` and then do branching logic in there to determine whether the tag should be the branch name, short commit hash, or `latest`.

We'd also need to decide if we want to push two tags in `cipublish` if we're on `master` (hash and `latest`), either in the script or by duplicating a check in the workflow config. That's kinda what I was trying to get away from in the old `.travis/deploy` script. Calling a simpler script twice with different arguments feels less complicated than a script that needs to have some nested conditional branches.

We could  definitely do it that way, but I feel that keeping the branching logic closer to the Actions context it conditions on and having `cipublish` only ever do one thing, publish with the provided tag, is easier to read and understand. What are your thoughts?

Reply:
> The catch here is that I'd like for it to be available to the workflow itself, and then have it passed along to `cipublish` as the `IMAGE_VERSION` variable. That lets us condition what kind of tag we want to push in the workflow instead of needing to pass the ref into `cipublish` and then do branching logic in there to determine whether the tag should be the branch name, short commit hash, or `latest`.

This makes sense 👍 .

> [...] That's kinda what I was trying to get away from in the old `.travis/deploy` script. Calling a simpler script twice with different arguments feels less complicated than a script that needs to have some nested conditional branches.

On the contrary, we've used conditional logic in `cipublish` [for things like](https://github.com/geotrellis/geotrellis-server/blob/01b01224545a36ad58731c6d1c19843212137656/scripts/cipublish#L20-L24) Scala artifact publishing.

> We could definitely do it that way, but I feel that keeping the branching logic closer to the Actions context it conditions on and having `cipublish` only ever do one thing, publish with the provided tag, is easier to read and understand.

While I agree that having `cipublish` do only one thing is easier to read, I disagree that it is easier to understand. `cipublish` is supposed to wrap the workflow so that I could run `cipublish` on my laptop, and everything would work. Right now, it wouldn't work, and understanding the workflow requires you to look in two places.

What are your thoughts on this?

```bash
if [[ -n "${GITHUB_SHA}" ]]; then
    GITHUB_SHA="${GITHUB_SHA:0:7}"
else
    GITHUB_SHA="$(git rev-parse --short HEAD)"
fi

if [[ -n "${GITHUB_REF}" ]]; then
    GITHUB_REF="${GITHUB_REF}"
else
    # We don't know where GitHub sources the value of GITHUB_REF. git-describe
    # is close enough to be valid for publishing.
    GITHUB_REF="$(git describe --all)"
fi
```

```bash
docker login -u ${QUAY_USER} -p ${QUAY_PASSWORD} quay.io

if [[ "${GITHUB_REF}" == "refs/heads/master" ]]; then
    # On master, use the current SHA and latest
    docker tag "raster-vision-${IMAGE_TYPE}" \
        "quay.io/azavea/raster-vision:${IMAGE_TYPE}-${GITHUB_SHA}"
    docker push "quay.io/azavea/raster-vision:${IMAGE_TYPE}-${GITHUB_SHA}"

    docker tag "raster-vision-${IMAGE_TYPE}" \
        "quay.io/azavea/raster-vision:${IMAGE_TYPE}-latest"
    docker push "quay.io/azavea/raster-vision:${IMAGE_TYPE}-latest"
else
    # For everything else, use the branch or tag name
    docker tag "raster-vision-${IMAGE_TYPE}" \
        "quay.io/azavea/raster-vision:${IMAGE_TYPE}-${GITHUB_REF##*/}"
    docker push "quay.io/azavea/raster-vision:${IMAGE_TYPE}-${GITHUB_REF##*/}"
fi
```

Reply:
I think that's an excellent approach and I appreciate you taking the time to write it out so I can see how it looks! That completely addresses my concerns about too much nested conditional logic, reads well, and is more in line with our typical approach. I'll work this into `cipublish`.